### PR TITLE
Game finish feaures

### PR DESCRIPTION
**Pull request by LiamMoMurray**
Games now finish when there is only one player left in a game.

here is the game before it is started
game before start

and the players in the game:
players before game start

game started, notice it changes to ACTIVE:
game start

and players after game start, teams move from next_pick to team_pick:
players after game start

below is players after the first round, both players that selected Liverpool are eliminated (is_Active = 0)
[i noticed after testing that Liverpool won this week, i had to change the date ranges to +1 day to fix this, working as intended now.]
players after round 1

next test was if both players lose, here both have Aston VIlla, so both are out:
both lose

because there is no players left, both stay in:
still in

after this round the player that picked Tottenham wins and the game is complete:
game complete

image

I am going to try get the new test endpoints to work only if the user is an admin.
and this game week according to here:
image

was Saturday to Monday, so the Newcastle-West Ham game was not included in the results, I might have to change the dates for rounds or add a check on if a team picked is playing within the round week.

### CODE REVIEW

**Review by Whomstt — APPROVED** [review]
Looks good to me. Good work.

File: game-service/src/main/java/com/lms/gameservice/controller/GameController.java [hunk 1]
```diff
@@ -224,6 +242,70 @@ public ResponseEntity<String> changeTeam(
...
+
+        } catch (Exception e) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
+                    .body("Error verifying the player: " + e.getMessage());
+        }
+    }
+
+    @GetMapping("/{game_id}/usedTeams")
+    public ResponseEntity<?> getUsedTeams(
+            @RequestHeader("Authorization") String authorisationHeader,
```
Comment (olanhealy) [inline game-service/src/main/java/com/lms/gameservice/controller/GameController.java:279]: misspelling

File: game-service/src/main/java/com/lms/gameservice/controller/GameController.java [hunk 2]
```diff
@@ -224,6 +242,70 @@ public ResponseEntity<String> changeTeam(
         }
     }
 
+    @GetMapping("/{game_id}/availableTeams")
+    public ResponseEntity<?> getAvailableTeams(
+            @RequestHeader("Authorization") String authorisationHeader,
```
Comment (olanhealy) [inline game-service/src/main/java/com/lms/gameservice/controller/GameController.java:247]: mispspelling

File: game-service/src/test/java/com/lms/gameservice/PaymentServiceClientTest.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package com.lms.gameservice.service;
```
Comment (olanhealy) [inline game-service/src/test/java/com/lms/gameservice/PaymentServiceClientTest.java:1]: Whats this class purpose?

File: game-service/src/main/java/com/lms/gameservice/service/PlayerService.java [hunk 4]
```diff
@@ -83,6 +83,31 @@ public void changeTeamPick(Player player, String newTeam){
...
+        player.setTeamsAvailable(availableTeams);
+
+        if(player.getTeamsUsed() == null){
+            player.setTeamsUsed(new ArrayList<>());
+        }
+        ArrayList<String> usedTeams = player.getTeamsUsed();
+        usedTeams.add(team);
+        player.setTeamsUsed(usedTeams);
+        player.setTeamPick(team);
+
```
Comment (olanhealy) [inline game-service/src/main/java/com/lms/gameservice/service/PlayerService.java:106]: V good like this

File: game-service/src/main/java/com/lms/gameservice/service/GameService.java [hunk 5]
```diff
@@ -180,18 +184,58 @@ public void updatePlayers(Game game) {
...
+
+            for (Player player : playersOut) { 
                 player.setActive(false);
                 playerRepository.save(player);
             }
         }
+
+    }
+
+    private void declareWinner(Game game, Player winner) {
```
Comment (olanhealy) [inline game-service/src/main/java/com/lms/gameservice/service/GameService.java:226]: good

File: game-service/src/main/java/com/lms/gameservice/controller/GameController.java [hunk 6]
```diff
@@ -281,31 +363,59 @@ private String extractUidFromMessage(String message) {
...
+     * 
+     */
+    @PostMapping("/{game_id}/gameStartTest")
+    public void startGame(@PathVariable("game_id") int gameId) {
+        
+        Game game = gameRepository.findById(gameId)
+                .orElseThrow(() -> new IllegalArgumentException("Game not found"));
+
+                gameService.startGame(game.getId());
+    }
```
Comment (olanhealy) [inline game-service/src/main/java/com/lms/gameservice/controller/GameController.java:420]: So will an admin have to set this? Or will this be automatically ran

**Review by olanhealy — CHANGES_REQUESTED** [review]
Some comments, mostly looks good to me great work

File: game-service/src/main/java/com/lms/gameservice/controller/GameController.java (hunk 1 above)
Comment (LiamMoMurray) [reply game-service/src/main/java/com/lms/gameservice/controller/GameController.java]: spelling is updated in latest commit

File: game-service/src/main/java/com/lms/gameservice/controller/GameController.java (hunk 2 above)
Comment (LiamMoMurray) [reply game-service/src/main/java/com/lms/gameservice/controller/GameController.java]: spelling is updated in latest commit

File: game-service/src/test/java/com/lms/gameservice/PaymentServiceClientTest.java (hunk 3 above)
Comment (LiamMoMurray) [reply game-service/src/test/java/com/lms/gameservice/PaymentServiceClientTest.java:1]: Callum made this, I think it is to pay when joining a game. I'll have a look cause there is no updates for the prize pot when people join and I don't think there is a check to see if a person has the balance to join.

File: game-service/src/main/java/com/lms/gameservice/controller/GameController.java (hunk 6 above)
Comment (LiamMoMurray) [reply game-service/src/main/java/com/lms/gameservice/controller/GameController.java:420]: no this should happen automatically from a similar function GameScheduler. I will make it so only admins can use the three test endpoints

**Review by olanhealy — APPROVED** [review]
Comments addressed, approving

**Pull request merged**